### PR TITLE
mgr / volume: refactor [sub]volume 

### PR DESCRIPTION
**Pull request by vshankar**
Based on top of PR #27594  (may be not the recent update). This carves [sub]volume into separate classes needed for building background delete support (and probably things like auto spawn MDSs on volume creation).

Keeping this as WIP, although it passed the test coverage PR #27856. Need to fill in some code/class/function commentary and reduce some redundant stat() calls which I happen to notice during the refactor.

Note that, we should aim to merge PR #27594 by all means before getting this in.

### CODE REVIEW

**Comment by batrick** [comment]
needs rebase

File: src/pybind/mgr/volumes/volume.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+import logging
+import orchestrator
+
+from path import SubvolumeSpec
+from subvolume import SubVolume
+
+log = logging.getLogger(__name__)
+
+class Volume(object):
+    def __init__(self, mgr, orchestrator):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/volume.py]: `orchestrator` is a python module. There is no need to keep that anywhere.

Comment (vshankar) [reply src/pybind/mgr/volumes/volume.py]: yep -- already fixed that w/ the rebase.

**Comment by vshankar** [comment]
@ajarr @batrick -- rebased and updated. please review.

File: src/pybind/mgr/volumes/subvolume.py [hunk 2]
```diff
@@ -228,11 +160,62 @@ def rmtree(root_path):
...
+
+    def get_group_path(self, spec, create_default=False):
+        path = spec.make_group_path()
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            if create_default and spec.is_default_group():
+                self._mkdir_p(path)
+                return path
+            return None
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/subvolume.py]: caller often raise an exception
```suggestion
            raise VolumeException(
                        -errno.ENOENT, "Subvolume group '{0}' not found".format(groupname))
```
does it make sense to move this into `get_group_path`?

File: src/pybind/mgr/volumes/exception.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class VolumeException(Exception):
+    def __init__(self, error_code, error_message):
+        super(VolumeException, self).__init__(error_code, error_message)
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/mgr/volumes/exception.py]: true -- but I kept that so as to use `args` (later) -- if its not really needed, I'll remove that on next update.

File: src/pybind/mgr/volumes/subvolume.py [hunk 4]
```diff
@@ -228,11 +160,62 @@ def rmtree(root_path):
...
+        """
+        grouppath = spec.make_group_path()
+        return self._list_dir(grouppath)
+
+    def get_subvolume_path(self, spec):
+        path = spec.make_fs_path()
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            return None
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/mgr/volumes/subvolume.py]: I had thought about doing that, but thought of keeping `get_subvolume_path()` exception free as it *might* have a use for some basic checks such "is there a subvolume in a given group?".

File: src/pybind/mgr/volumes/volume.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    def volume_exists(self, volname):
+        return self.get_fs(volname) is not None
+
+    def volume_exception_to_retval(self, ve):
+        """
+        return a tuple representation from a volume exception
+        """
+        return ve.to_tuple()
+
+    def create_pool(self, pool_name, pg_num, pg_num_min=None, pg_autoscale_factor=None):
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/mgr/volumes/volume.py]: thank you! I'll take a look at that (this is pretty new to me)

File: src/pybind/mgr/volumes/subvolume.py (hunk 2 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/subvolume.py]: ditto as per `get_subvolume_path()`

File: src/pybind/mgr/volumes/subvolume.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/volumes/subvolume.py]: Got that, thanks!

(just fyi, the pythonic way would be **`in`** and raise a `KeyError`, if the key is not found.:

```
>>> my_dict = {'key': 'value'}

>>> my_dict['unknown']
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-17-3338a505bacc> in <module>()
----> 1 my_dict['unknown']

KeyError: 'unknown'

>>> 'key' in my_dict
True
```

**Comment by batrick** [comment]
@sebastian-philipp any other comments?

@vshankar I'm cool with this with Ramana's review.

File: src/pybind/mgr/volumes/volume.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            return []
+        return [mds['name'] for mds in fs['mdsmap']['info'].values()]
+
+    def volume_exists(self, volname):
+        return self.get_fs(volname) is not None
+
+    def create_pool(self, pool_name, pg_num, pg_num_min=None, pg_autoscale_factor=None):
+        # create the given pool
+        command = {'prefix': 'osd pool create', 'pool': pool_name, 'pg_num': pg_num}
+        if pg_num_min is not None:
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: if not pg_num_min:

File: src/pybind/mgr/volumes/volume.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            command['pg_num_min'] = pg_num_min
+        r, outb, outs = self.mgr.mon_command(command)
+        if r != 0:
+            return r, outb, outs
+
+        # set pg autoscale if needed
+        if pg_autoscale_factor is not None:
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: if not pg_autoscale_factor:

File: src/pybind/mgr/volumes/volume.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        # create pools
+        r, outs, outb = self.create_pool(metadata_pool, 16, pg_num_min=16, pg_autoscale_factor=4.0)
+        if r != 0:
+            return r, outb, outs
+        r, outb, outs = self.create_pool(data_pool, 8)
+        if r != 0:
+            return r, outb, outs
+        # create filesystem
+        r, outb, outs = self.create_filesystem(volname, metadata_pool, data_pool)
+        if r != 0:
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: Can we add this log message, as it was before,
```
self.log.error("Filesystem creation error: {0} {1} {2}".format(r, outb, outs))
```

File: src/pybind/mgr/volumes/volume.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        spec.name = fs_name
+        try:
+            completion = self.mgr.add_stateless_service("mds", spec)
+            self.mgr._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except (ImportError, orchestrator.OrchestratorError):
+            return 0, "", "Volume created successfully (no MDS daemons created)"
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: Can we add this log message as before,
self.log.exception("Failed to create MDS daemons")

File: src/pybind/mgr/volumes/volume.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+            pass
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            return -errno.EINVAL, "", str(e)
+
+        # In case orchestrator didn't tear down MDS daemons cleanly, or
+        # there was no orchestrator, we force the daemons down.
+        if self.volume_exists(volname):
+            r, outb, outs = self.teardown_mds(volname)
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: The naming of the method is confusing. Here you don't teardown MDS, you set the mds daemons down. Please rename this method.

File: src/pybind/mgr/volumes/volume.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            orchestrator.raise_if_exception(completion)
+        except (ImportError, orchestrator.OrchestratorError):
+            return 0, "", "Volume created successfully (no MDS daemons created)"
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
+            return -errno.EINVAL, "", str(e)
+        return 0, "", ""
+
+    def teardown_mds(self, fs_name):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: rename this method. you don't teardown_mds here.

File: src/pybind/mgr/volumes/volume.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        r, outb, outs = self.create_filesystem(volname, metadata_pool, data_pool)
+        if r != 0:
+            return r, outb, outs
+        # create mds
+        return self.create_mds(volname)
+
+    def delete_volume(self, volname):
+        """
+        """
+        # Tear down MDS daemons
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: These steps should be in teardown_mds method instead

File: src/pybind/mgr/volumes/volume.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+        # set pg autoscale if needed
+        if pg_autoscale_factor is not None:
+            command = {'prefix': 'osd pool set', 'pool': pool_name, 'var': 'pg_autoscale_bias',
+                       'val': str(pg_autoscale_factor)}
+            r, outb, outs = self.mgr.mon_command(command)
+        return r, outb, outs
+
+    def remove_pool(self, pool_name):
+        command = {'prefix': 'osd pool rm', 'pool': pool_name, 'pool2': pool_name,
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: Do you know why we remove 'pool2' which is same as 'pool'?

File: src/pybind/mgr/volumes/volume.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+            r, outb, outs = self.teardown_mds(volname)
+            if r != 0:
+                return r, outb, outs
+            r, outb, outs = self.remove_filesystem(volname)
+            if r != 0:
+                return r, outb, outs
+        else:
+            log.warning("Filesystem already gone for volume '{0}'".format(volname))
+            # log here...
+            pass
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: Is pass required here?

Comment (vshankar) [reply src/pybind/mgr/volumes/volume.py]: leftover from debug.. thanks for catching.

File: src/pybind/mgr/volumes/module.py [hunk 15]
```diff
@@ -11,6 +11,7 @@
 import orchestrator
 
 from .fs.subvolume import SubvolumePath, SubvolumeClient
+import volume
```
Comment (ajarr) [inline src/pybind/mgr/volumes/module.py]: I prefer if this module was fs.volume, like fs.subvolume. We might in the future have RBD volumes module.

File: src/pybind/mgr/volumes/module.py [hunk 16]
```diff
@@ -176,107 +178,20 @@ def handle_command(self, inbuf, cmd):
...
-    def _volume_get_mds_daemon_names(self, vol_name):
-        fs = self._volume_get_fs(vol_name)
-        if fs is None:
-            return []
+        return self.vc.create_volume(vol_id, size)
 
-        return [i['name'] for i in fs['mdsmap']['info'].values()]
+    def _cmd_fs_volume_rm(self, inbuf, cmd):
+        vol_name = cmd['vol_name']
+        return self.vc. delete_volume(vol_name)
```
Comment (ajarr) [inline src/pybind/mgr/volumes/module.py]: there is an incorrect space between 'vc.' and 'delete_volume'

File: src/pybind/mgr/volumes/volume.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+        metadata_pool, data_pool = self.gen_pool_names(volname)
+        r, outb, outs = self.remove_pool(metadata_pool)
+        if r != 0:
+            return r, outb, outs
+        return self.remove_pool(data_pool)
+
+    def volume_list(self):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: maybe rename as list_volumes() for sake of uniformity with other method names?

File: src/pybind/mgr/volumes/pathspec.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        self.pool_ns_prefix = pool_ns_prefix if pool_ns_prefix else SubvolumeSpec.DEFAULT_NS_PREFIX
+
+    def is_default_group(self):
+        """
+        Is the group the default group?
+        """
+        return self.groupid == SubvolumeSpec.NO_GROUP_NAME
+
+    @property
+    def fs_path(self):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/pathspec.py]: s/fs_path/subvol_path/ ?

File: src/pybind/mgr/volumes/pathspec.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+        return self.subvolume_prefix
+
+    @property
+    def trash_dir(self):
+        """
+        return the trash directory path
+        """
+        return os.path.join(self.subvolume_prefix, "_deleting")
+
+    def make_subvol_snap_path(self, snapdir, snapname):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/pathspec.py]: @property?

File: src/pybind/mgr/volumes/pathspec.py [hunk 20]
```diff
@@ -0,0 +1,87 @@
+import os
+
+class SubvolumeSpec(object):
+    """
+    Specification of a subvolume, identified by (subvolume-id, group-id) tuple. Add fields as
```
Comment (ajarr) [inline src/pybind/mgr/volumes/pathspec.py]: "Add fields as required...". What does that mean?

File: src/pybind/mgr/volumes/pathspec.py (hunk 19 above)
Comment (ajarr) [inline src/pybind/mgr/volumes/pathspec.py]: Can we make add @property here as well?

File: src/pybind/mgr/volumes/pathspec.py [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+        """
+        return os.path.join(self.subvolume_prefix, "_deleting")
+
+    def make_subvol_snap_path(self, snapdir, snapname):
+        """
+        return the subvolume snapshot path for a given snapshot name
+        """
+        return os.path.join(self.fs_path, snapdir, snapname)
+
+    def make_group_snap_path(self, snapdir, snapname):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/pathspec.py]: Can we add the property decorator here as wekk, and rename the method as group_snap_path()?

File: src/pybind/mgr/volumes/pathspec.py (hunk 19 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/pathspec.py]: can functions w/ params be turned into a property?

File: src/pybind/mgr/volumes/pathspec.py (hunk 21 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/pathspec.py]: ditto as above.

File: src/pybind/mgr/volumes/pathspec.py (hunk 19 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/pathspec.py]: you're right. I don't think so. It's the setter that can have params.

File: src/pybind/mgr/volumes/volume.py [hunk 22]
```diff
@@ -1,18 +1,21 @@
...
 
 class VolumeClient(object):
     def __init__(self, mgr):
-        """
-        """
         self.mgr = mgr
 
     def gen_pool_names(self, volname):
         """
+        return metadata and data pool name (from a filesystem/volume name) as a tuple
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: This can be in the initial, "carve out volume" commit

File: src/pybind/mgr/volumes/volume.py [hunk 23]
```diff
@@ -92,8 +95,6 @@ def teardown_mds(self, fs_name):
     ### volume operations -- create, rm, ls
 
     def create_volume(self, volname, size=None):
-        """
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: This can be in the initial, "carve out volume" commit

File: src/pybind/mgr/volumes/volume.py [hunk 24]
```diff
@@ -110,8 +111,6 @@ def create_volume(self, volname, size=None):
         return self.create_mds(volname)
 
     def delete_volume(self, volname):
-        """
-        """
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: This can be in the initial, "carve out volume" commit

File: src/pybind/mgr/volumes/module.py [hunk 25]
```diff
@@ -200,246 +198,81 @@ def _cmd_fs_subvolumegroup_create(self, inbuf, cmd):
...
-                           "Subvolume '{0}' not found, cannot remove it".format(sub_name)
-            svc.purge_subvolume(svp)
-
-        # TODO: purge subvolume asynchronously
-        # self._background_jobs.put(PurgeJob(vol_fscid, svp))
-
-        return 0, "", ""
+        return self.vc.remove_subvolume(vol_name, sub_name, group_name, force)
+
+    def _cmd_fs_subvolume_ls(self, inbuf, cmd):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/module.py]: Let's not introduce listing of subvolumes in this commit. It should be a separate commit. Last time I introduced listing, Patrick had certain issues around this,
https://github.com/ceph/ceph/pull/27594#discussion_r280285238

File: src/pybind/mgr/volumes/subvolume.py [hunk 26]
```diff
@@ -0,0 +1,275 @@
+import os
```
Comment (ajarr) [inline src/pybind/mgr/volumes/subvolume.py]: I prefer that this module be in the previous location, fs/subvolume.py.

File: src/pybind/mgr/volumes/subvolume.py [hunk 27]
```diff
@@ -0,0 +1,275 @@
...
+                        self.fs.unlink(d_full)
+
+                d = self.fs.readdir(dir_handle)
+            self.fs.closedir(dir_handle)
+            self.fs.rmdir(root_path)
+
+        trashpath = spec.trash_path
+        rmtree(trashpath)
+
+    def list_subvolumes(self, spec):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/subvolume.py]: Let's not introduce listing of subvolumes in this commit. It should be a separate commit. Last time I introduced listing, Patrick had certain issues around this,
#27594 (comment)

File: src/pybind/mgr/volumes/subvolume.py [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+
+    def create_group(self, spec, mode=0o755):
+        path = spec.group_path
+        self._mkdir_p(path, mode)
+
+    def remove_group(self, spec):
+        path = spec.group_path
+        self.fs.rmdir(path)
+
+    def list_groups(self):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/subvolume.py]: Let's not introduce listing of subvolumes in this commit. It should be a separate commit. Last time I introduced listing, Patrick had certain issues around this,
#27594 (comment)

File: src/pybind/mgr/volumes/pathspec.py [hunk 29]
```diff
@@ -0,0 +1,87 @@
+import os
```
Comment (ajarr) [inline src/pybind/mgr/volumes/pathspec.py]: I prefer that this module be at pybind/mgr/volumes/fs/ folder and named something else, e.g., subvolspec.py

File: src/pybind/mgr/volumes/volume.py [hunk 30]
```diff
@@ -0,0 +1,151 @@
+import json
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: I prefer this module's path was,
src/pybind/mgr/volumes/fs/volume.py

File: src/pybind/mgr/volumes/volume.py [hunk 31]
```diff
@@ -149,3 +148,207 @@ def volume_list(self):
...
+                sv.remove_group(spec)
+            except cephfs.ObjectNotFound:
+                if force:
+                    return 0, "", ""
+                else:
+                    return -errno.ENOENT, "", \
+                        "Subvolume group '{0}' not found".format(groupname)
+        return 0, "", ""
+
+    def list_subvolume_groups(self, volname):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/volume.py]: Let's not introduce listing in this commit. Let it be a new commit

File: src/pybind/mgr/volumes/pathspec.py (hunk 18 above)
Comment (ajarr) [inline src/pybind/mgr/volumes/pathspec.py]: rename as subvol_path to be more clear?

File: src/pybind/mgr/volumes/subvolume.py [hunk 32]
```diff
@@ -0,0 +1,275 @@
...
+        snappath = spec.make_group_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        return self._snapshot_delete(snappath)
+
+    def connect(self):
+        log.debug("Connecting to cephfs...")
+        self.fs = cephfs.LibCephFS(rados_inst=self.rados)
+        log.debug("CephFS initializing...")
+        self.fs.init()
+        log.debug("CephFS mounting...")
+        self.fs.mount(filesystem_name=self.fs_name)
```
Comment (ajarr) [inline src/pybind/mgr/volumes/subvolume.py]: See, https://github.com/ceph/ceph/pull/28380/commits/a203e577098d82a6d0dea86757c9e83dccdba78d

Comment (vshankar) [reply src/pybind/mgr/volumes/subvolume.py]: already did that when rebasing yesterday (and another similar change in `setxattr`)

File: src/pybind/mgr/volumes/volume.py [hunk 33]
```diff
@@ -1,18 +1,21 @@
...
 import orchestrator
 
+from pathspec import SubvolumeSpec
+from subvolume import SubVolume
+
 log = logging.getLogger(__name__)
 
 class VolumeClient(object):
     def __init__(self, mgr):
-        """
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/mgr/volumes/volume.py]: oh -- I thought I had already done that :/

will fix.

File: src/pybind/mgr/volumes/pathspec.py (hunk 20 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/pathspec.py]: in the sense, add fields that identify a (unique) subvolume.

File: src/pybind/mgr/volumes/module.py (hunk 25 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/module.py]: are you proposing list/get to be in a separate commit or not to have it at all? according to the comment you mention, it seems to me to introduce it later once there is access control for those.

File: src/pybind/mgr/volumes/subvolume.py (hunk 28 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/subvolume.py]: ditto as per subvolume list.

File: src/pybind/mgr/volumes/volume.py (hunk 11 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/volume.py]: calling it `set_mds_down`

File: src/pybind/mgr/volumes/volume.py (hunk 13 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/volume.py]: I haven't looked into this (yet) -- just move the relevant parts from `module` to this source.

File: src/pybind/mgr/volumes/volume.py (hunk 7 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/volume.py]: Sorry, this should be if pg_autoscale_factor:

File: src/pybind/mgr/volumes/volume.py (hunk 6 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/volume.py]: This should be `if pg_num_min:`

Comment (vshankar) [reply src/pybind/mgr/volumes/volume.py]: already took care of that ;)

File: src/pybind/mgr/volumes/volume.py (hunk 7 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/volume.py]: already took care of that ;)

File: src/pybind/mgr/volumes/module.py (hunk 25 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/module.py]: I think it should be a separate commit at the least. Then we can check with Patrick.

Comment (vshankar) [reply src/pybind/mgr/volumes/module.py]: I'll vote for introducing that as a separate pr. @batrick ?

Comment (batrick) [reply src/pybind/mgr/volumes/module.py]: separate PR sounds good

**Comment by vshankar** [comment]
@ajarr fixed, rebased and updated.

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 34]
```diff
@@ -4,38 +4,18 @@
...
 
-import errno
 import logging
 import os
+import errno
 
 import cephfs
-import rados
 
+from subvolspec import SubvolumeSpec
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/subvolume.py]: s/subvolspec/.subvolspec/

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 35]
```diff
@@ -4,38 +4,18 @@
...
+from exception import VolumeException
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/subvolume.py]: s/exception/.exception/

There is a python inbuilt module called 'exceptions'.  I think it'll be better if you rename the 'exception' module as 'volume_exception'.

File: src/pybind/mgr/volumes/fs/volume.py [hunk 36]
```diff
@@ -0,0 +1,352 @@
+import json
+import errno
+import cephfs
+import logging
+import orchestrator
+
+from subvolspec import SubvolumeSpec
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/volume.py]: s/subvolspec/.subvolspec/

File: src/pybind/mgr/volumes/fs/volume.py [hunk 37]
```diff
@@ -0,0 +1,352 @@
...
+from subvolume import SubVolume
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/volume.py]: s/subvolume/.subvolume/

File: src/pybind/mgr/volumes/fs/volume.py [hunk 38]
```diff
@@ -0,0 +1,352 @@
...
+from exception import VolumeException
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/volume.py]: s/exception/.exception/

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 39]
```diff
@@ -252,56 +202,63 @@ def _get_ancestor_xattr(self, path, attr):
...
         """
-        snapshot_path = self._snapshot_path(dir_path, snapshot_name)
-        self.fs.stat(snapshot_path)
-        self.fs.rmdir(snapshot_path)
+        self.fs.stat(snappath)
+        self.fs.rmdir(snappath)
 
-    def create_subvolume_snapshot(self, subvolume_path, snapshot_name, mode=0o755):
-        return self._snapshot_create(self._subvolume_path(subvolume_path), snapshot_name, mode)
+    def create_subvolume_snapshot(self, spec, snapname, mode=0o755):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/subvolume.py]: Why are subvolume snapshot create and rm methods under 'group operations' category?

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 40]
```diff
@@ -252,56 +202,63 @@ def _get_ancestor_xattr(self, path, attr):
...
+        snappath = spec.make_group_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        self._snapshot_create(snappath, mode)
 
-    def delete_group_snapshot(self, group_id, snapshot_name):
-        return self._snapshot_delete(self._group_path(group_id), snapshot_name)
+    def remove_group_snapshot(self, spec, snapname):
+        snappath = spec.make_group_snap_path(self.rados.conf_get('client_snapdir'), snapname)
+        return self._snapshot_delete(snappath)
+
+    def connect(self):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/subvolume.py]: Not sure why you shuffled connect() and the below methods to the end of the module?

File: src/pybind/mgr/volumes/fs/subvolume.py (hunk 34 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/subvolume.py]: Without this change I hit the error that the subvolspec module couldn't be found

File: src/pybind/mgr/volumes/fs/volume.py (hunk 38 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/fs/volume.py]: doesn't python look for the module in the current directory first? is there a need to explicitly mention with `.`?

File: src/pybind/mgr/volumes/fs/subvolume.py (hunk 40 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/fs/subvolume.py]: the internal context manager routine overrides (and helper routines) were right in between the subvolume logic -- https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/fs/subvolume.py#L92 (between `_group_path()` and `_mkdir_p()`).

**Comment by vshankar** [comment]
@ajarr regarding the module import error you observe, how were you testing this pr?

This PR passes tests in teuthology (w/ purge queue changes + volume python tests) -- http://pulpito.ceph.com/vshankar-2019-06-11_06:57:05-fs-wip-subvol-purge-queue-distro-basic-mira/

**Comment by ajarr** [comment]
> @ajarr regarding the module import error you observe, how were you testing this pr?
> 
> This PR passes tests in teuthology (w/ purge queue changes + volume python tests) -- http://pulpito.ceph.com/vshankar-2019-06-11_06:57:05-fs-wip-subvol-purge-queue-distro-basic-mira/

I tested the PR with a vstart dev cluster in py3 environment. Does teuthology also test in py3 environment? I thought it did only py2 tests.

File: src/pybind/mgr/volumes/fs/volume.py [hunk 41]
```diff
@@ -0,0 +1,153 @@
+import json
+import errno
+import logging
+import orchestrator
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/volume.py]: Introduce a blank line between built-in modules, and third-party modules.
https://pep8.org/#imports

I see at least this convention being followed in other python mgr modules

File: src/pybind/mgr/volumes/fs/volume.py (hunk 38 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/volume.py]: If you want to do relative imports, as you intend to do here, you need to precede the module with a '.' . See other examples,
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/ansible/module.py#L19

If you want to do absolute imports, then you need to use the full path of the module from project's root folder.
https://realpython.com/absolute-vs-relative-python-imports/

File: src/pybind/mgr/volumes/fs/volume.py [hunk 42]
```diff
@@ -0,0 +1,352 @@
...
+    def create_subvolume(self, volname, subvolname, groupname, size):
+        ret = 0, "", ""
+        try:
+            if not self.volume_exists(volname):
+                raise VolumeException(
+                    -errno.ENOENT, "Volume '{0}' not found, create it with `ceph fs " \
+                    "volume create` before trying to create subvolumes".format(volname))
+            with SubVolume(self.mgr, fs_name=volname) as sv:
+                spec = SubvolumeSpec(subvolname, groupname)
+                if not sv.get_group_path(spec, create_default=True):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/volume.py]: I don't follow, why you need 'create_default', here and elsewhere? It creates the default subvolume group? The create_subvolume() will create the default subvolume group.

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/volume.py]: That *might* have been necessary for supporting subvolume/group list commands. Since those are not in this PR now, I need to recheck if this is still required. WIll get to this when I address comments.

File: src/pybind/mgr/volumes/fs/subvolume.py (hunk 40 above)
Comment (ajarr) [reply src/pybind/mgr/volumes/fs/subvolume.py]: OK

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 43]
```diff
@@ -133,80 +55,80 @@ def _mkdir_p(self, path, mode=0o755):
...
                 self.fs.mkdir(subpath, mode)
 
-    def create_group(self, group_id, mode=0o755):
-        path = self._group_path(group_id)
-        self._mkdir_p(path, mode)
-
-    def delete_group(self, group_id):
-        path = self._group_path(group_id)
-        self.fs.rmdir(path)
+    def _list_dir(self, dir_path):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/subvolume.py]: I don't think we need this method in this PR

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/subvolume.py]: yeh -- wil remove.

**Comment by vshankar** [comment]
> I tested the PR with a vstart dev cluster in py3 environment. Does teuthology also test in py3 environment? I thought it did only py2 tests.

Not really sure if teuthology only uses py2. So, did py3 make the use of relative imports prefixing with `.` mandatory?

**Comment by ktdreyer** [comment]
@batrick I think we wanted to cherry-pick this for nautilus for Ceph CSI. I don't see a tracker.ceph.com ticket for that - want to file one?

File: src/pybind/mgr/volumes/fs/subvolume.py (hunk 35 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/fs/subvolume.py]: it's pretty normal to name it as `exception.py` imo.

**Comment by vshankar** [comment]
@ktdreyer http://tracker.ceph.com/issues/39969http://tracker.ceph.com/issues/39969

I'll include that in the commit msg on next push.

**Comment by ajarr** [comment]
> > I tested the PR with a vstart dev cluster in py3 environment. Does teuthology also test in py3 environment? I thought it did only py2 tests.
> 
> Not really sure if teuthology only uses py2. So, did py3 make the use of relative imports prefixing with `.` mandatory?

Good question. https://stackoverflow.com/questions/30249119/difference-in-package-importing-between-python-2-7-and-3-4

File: src/pybind/mgr/volumes/fs/subvolume.py (hunk 35 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/fs/subvolume.py]: in the sense, it's common to stash various custom exception classes into one exception source.

**Comment by batrick** [comment]
needs rebased

File: src/pybind/mgr/volumes/fs/subvolume.py (hunk 39 above)
Comment (vshankar) [reply src/pybind/mgr/volumes/fs/subvolume.py]: that's due to fall through from "groups" comment.

File: src/pybind/mgr/volumes/fs/subvolume.py [hunk 44]
```diff
@@ -54,12 +56,18 @@ def _mkdir_p(self, path, mode=0o755):
...
 
     def _list_dir(self, dir_path):
         """
         Returns a list of directory names within a directory
         """
-        dir_handle = self.fs.opendir(dir_path)
+        try:
+            dir_handle = self.fs.opendir(dir_path)
+        except cephfs.Error as e:
+            raise VolumeError(e.args[0], e.args[1])
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/subvolume.py]: s/VolumeError/VolumeException/ ?

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/subvolume.py]: already fixed...

**Comment by batrick** [comment]
http://pulpito.ceph.com/?branch=wip-pdonnell-testing-20190614.175650
